modules/auxiliary/scanner/http/options.rb

```diff
@@ -5,69 +5,80 @@
 
 class MetasploitModule < Msf::Auxiliary
 
-  # Exploit mixins should be called first
   include Msf::Exploit::Remote::HttpClient
   include Msf::Auxiliary::WmapScanServer
-  # Scanner mixin should be near last
   include Msf::Auxiliary::Scanner
   include Msf::Auxiliary::Report
 
   def initialize
     super(
-      'Name'        => 'HTTP Options Detection',
+      'Name' => 'HTTP Options Detection',
       'Description' => 'Display available HTTP options for each system',
-      'Author'       => ['CG'],
-      'License'     => MSF_LICENSE,
-      'References' =>
+      'Author' => ['CG'],
+      'License' => MSF_LICENSE,
+      'References' => [
+        ['CVE', '2005-3398'], # HTTP Trace related
+        ['CVE', '2005-3498'], # HTTP Trace related
+        ['OSVDB', '877'],
+        ['BID', '11604'],
+        ['BID', '9506'],
+        ['BID', '9561'],
+        ['URL', 'https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/OPTIONS']
+      ]
+    )
+    register_options(
       [
-        [ 'CVE', '2005-3398'],
-        [ 'CVE', '2005-3498'],
-        [ 'OSVDB', '877'],
-        [ 'BID', '11604'],
-        [ 'BID', '9506'],
-        [ 'BID', '9561']
+        OptString.new('TARGETURI', [true, 'URI to test', '/']),
       ]
     )
   end
 
   def run_host(target_host)
+    res = send_request_cgi({
+      'uri' => datastore['TARGETURI'],
+      'method' => 'OPTIONS'
+    })
+    return if !res
+
+    # patch so that tomcat which responds in HTTP will work as well
+    # https://github.com/rapid7/metasploit-framework/issues/12557#issuecomment-552263162
+    # https://stackoverflow.com/questions/23886941/http-status-405-jsps-only-permit-get-post-or-head
+    if res.body =~ %r{<h1>HTTP Status 405 - JSPs only permit ([^<]*)</h1>}
+      res.headers['Allow'] = ::Regexp.last_match(1).sub(' or ', ' ').gsub(' ', ', ')
+    end
 
-    begin
-      res = send_request_raw({
-        'version'      => '1.0',
-        'uri'          => '/',
-        'method'       => 'OPTIONS'
-      }, 10)
+    unless res.headers['Allow']
+      vprint_error("#{target_host} missing Allow header")
+      return
+    end
 
-      if (res and res.headers['Allow'])
-        print_good("#{target_host} allows #{res.headers['Allow']} methods")
+    allowed_methods = res.headers['Allow']
 
-        report_note(
-          :host	=> target_host,
-          :proto => 'tcp',
-          :sname => (ssl ? 'https' : 'http'),
-          :port	=> rport,
-          :type	=> 'HTTP_OPTIONS',
-          :data	=> res.headers['Allow']
-        )
+    print_good("#{target_host} allows #{allowed_methods} methods")
```

Comment:
Is this not an array of items normally? Your edge case seems to be converting this into a string but I'm not sure this code will work in normal circumstances.

Reply:
during testing I also added `puts res.headers['Allow'].class` and received `String` in the output, so I believe everything is a string that I'm dealing with.

Reply:
Thanks that is sufficient for me, will mark this as resolved, don't think this should be an issue.